nano_graphrag/_storage.py

```diff
@@ -134,85 +135,85 @@ def __post_init__(self):
         self.M = hnsw_params.get("M", self.M)
         self.max_elements = hnsw_params.get("max_elements", self.max_elements)
         self.ef_search = hnsw_params.get("ef_search", self.ef_search)
         self.num_threads = hnsw_params.get("num_threads", self.num_threads)
+        self._index = hnswlib.Index(space='cosine', dim=self.embedding_func.embedding_dim)
 
         if os.path.exists(self._index_file_name) and os.path.exists(self._metadata_file_name):
-            self._index = hnswlib.Index(space='cosine', dim=self.embedding_func.embedding_dim)
             self._index.load_index(self._index_file_name, max_elements=self.max_elements)
             with open(self._metadata_file_name, 'rb') as f:
                 self._metadata, self._current_elements = pickle.load(f)
             logger.info(f"Loaded existing index for {self.namespace} with {self._current_elements} elements")
         else:
-            self._index = hnswlib.Index(space='cosine', dim=self.embedding_func.embedding_dim)
             self._index.init_index(
                 max_elements=self.max_elements,
                 ef_construction=self.ef_construction,
                 M=self.M
             )
             self._index.set_ef(self.ef_search)
+            self._metadata = {}
+            self._current_elements = 0
             logger.info(f"Created new index for {self.namespace}")
 
-    async def upsert(self, data: dict[str, dict]):
+    async def upsert(self, data: dict[str, dict]) -> np.ndarray:
         logger.info(f"Inserting {len(data)} vectors to {self.namespace}")
         if not data:
-            raise ValueError("Attempting to insert empty data to vector DB")
+            logger.warning("You insert an empty data to vector DB")
+            return []
 
         if self._current_elements + len(data) > self.max_elements:
             raise ValueError(f"Cannot insert {len(data)} elements. Current: {self._current_elements}, Max: {self.max_elements}")
 
-        contents = [v["content"] for v in data.values()]
-        batches = [
-            contents[i : i + self._max_batch_size]
-            for i in range(0, len(contents), self._max_batch_size)
+        list_data = [
+            {
+                "id": k,
+                **{k1: v1 for k1, v1 in v.items() if k1 in self.meta_fields},
+            }
+            for k, v in data.items()
         ]
-        embeddings_list = await asyncio.gather(
-            *[self.embedding_func(batch) for batch in batches]
+        contents = [v["content"] for v in data.values()]
+        batch_size = min(self._embedding_batch_num, len(contents))
+        embeddings = np.concatenate(await asyncio.gather(
+            *[self.embedding_func(contents[i:i + batch_size]) 
+            for i in range(0, len(contents), batch_size)]
+        ))
+
+        ids = np.fromiter(
+            (xxhash.xxh32_intdigest(d["id"].encode()) for d in list_data), 
```

Comment:
Could we done this `str` to `int` in built-in `hashlib`?
Something like: `int(md5(d["id"].encode()).hexdigest(), 16)`